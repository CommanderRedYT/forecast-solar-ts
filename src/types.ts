export interface ForecastSolarOptions {
    azimuth: number;
    declination: number;
    kwp: number;
    latitude: number;
    longitude: number;
    apiKey?: string;
    damping?: number;
    dampingMorning?: number;
    dampingEvening?: number;
    horizon?: string;
    inverter?: number;
    baseEndpoint?: string;
}

export interface ForecastSolarData {
    [key: string]: number;
}
