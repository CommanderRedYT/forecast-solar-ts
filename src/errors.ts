/* eslint-disable max-classes-per-file */
export class ForecastSolarError extends Error {
    constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}

export class ForecastSolarConfigError extends ForecastSolarError {
    constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}

export class ForecastSolarConnectionError extends ForecastSolarError {
    constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}

export class ForecastSolarRequestError extends ForecastSolarError {
    constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}

export class ForecastSolarRateLimitError extends ForecastSolarError {
    constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}

export class ForecastSolarAuthenticationError extends ForecastSolarError {
    constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}
