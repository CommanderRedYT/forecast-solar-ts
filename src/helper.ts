import moment, { Moment } from 'moment';
import { ForecastSolarData } from './types';

/*
def _timed_value(at: datetime, data: dict[datetime, int]) -> int | None:
    """Return the value for a specific time."""
    value = None
    for timestamp, cur_value in data.items():
        if timestamp > at:
            return value
        value = cur_value

    return None

    def _interval_value_sum(
    interval_begin: datetime, interval_end: datetime, data: dict[datetime, int]
) -> int:
    """Return the sum of values in interval."""
    total = 0

    for timestamp, wh in data.items():
        # Skip all until this hour
        if timestamp < interval_begin:
            continue

        if timestamp >= interval_end:
            break

        total += wh

    return total
 */

export function timedValue(at: Moment, data: ForecastSolarData): number | null {
    let value: number | null = null;
    for (const [timestamp, curValue] of Object.entries(data)) {
        const timestampMoment = moment(timestamp);

        if (timestampMoment > at) {
            return value;
        }
        value = curValue;
    }

    return null;
}

export function intervalValueSum(intervalBegin: Moment, intervalEnd: Moment, data: ForecastSolarData): number {
    let total = 0;

    for (const [timestamp, wh] of Object.entries(data)) {
        const timestampMoment = moment(timestamp);

        if (timestampMoment < intervalBegin) {
            continue;
        }

        if (timestampMoment >= intervalEnd) {
            break;
        }

        total += wh;
    }

    return total;
}
