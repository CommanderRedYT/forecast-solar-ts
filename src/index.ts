import ForecastSolar from './forecastSolar';

export { ForecastSolar };

export type { ForecastSolarOptions } from './types';

export default ForecastSolar;
