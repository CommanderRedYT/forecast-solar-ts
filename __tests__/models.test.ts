import { MockServer } from 'jest-mock-server';
import moment from 'moment';
import { ForecastSolar, ForecastSolarOptions } from '../src';
import { loadFixtureJSON } from './utils';
import { AccountType } from '../src/models';
import {ForecastSolarError} from "../src/errors";

const clientOptions: ForecastSolarOptions = {
    latitude: 48.21,
    longitude: 16.36,
    azimuth: 180,
    declination: 23.44,
    kwp: 5,
};

const subscribedClientOptions: ForecastSolarOptions = {
    apiKey: 'myapikey',
    latitude: 52.16,
    longitude: 4.47,
    azimuth: 10,
    declination: 20,
    kwp: 2.160,
    dampingMorning: 0,
    dampingEvening: 0,
    horizon: '0,0,0,10,10,20,20,30,30',
    inverter: 1.300,
};

describe('Test estimated forecast', () => {
    const mockServer: MockServer = new MockServer();

    beforeAll(async () => {
        await mockServer.start();
    });

    afterAll(async () => {
        await mockServer.stop();
    });

    beforeEach(() => {
        mockServer.reset();
    });

    it('should return a forecast', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-06-11T12:00:00+02:00'));

        const route = mockServer.get('/estimate/48.21/16.36/23.44/180/5').mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Remaining', '99');
            ctx.body = loadFixtureJSON('forecast.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        const forecast = await instance.estimate();

        expect(forecast).toMatchSnapshot('snapshot1');
        expect(forecast.timezone).toBe('Europe/Vienna');
        expect(forecast.accountType).toBe(AccountType.Public);

        expect(forecast.energyProductionToday).toBe(2991);
        expect(forecast.energyProductionTomorrow).toBe(2602);

        expect(forecast.powerProductionNow).toBe(289);
        expect(forecast.energyProductionTodayRemaining).toBe(2059);
        expect(forecast.energyCurrentHour).toBe(270);

        expect(forecast.powerHighestPeakTimeToday).toBeInstanceOf(moment);
        expect(forecast.powerHighestPeakTimeToday).toStrictEqual(moment('2024-06-11T14:00:00+02:00'));

        expect(forecast.powerHighestPeakTimeTomorrow).toBeInstanceOf(moment);
        expect(forecast.powerHighestPeakTimeTomorrow).toStrictEqual(moment('2024-06-12T13:00:00+02:00'));

        expect(forecast.sumEnergyProduction(1)).toBe(312);
        expect(forecast.sumEnergyProduction(6)).toBe(1629);
        expect(forecast.sumEnergyProduction(12)).toBe(1789);
        expect(forecast.sumEnergyProduction(24)).toBe(2893);

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should return a forecast with subscription', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-04-27T07:00:00+02:00'));

        const route = mockServer.get(`/${subscribedClientOptions.apiKey}/estimate/52.16/4.47/20/10/2.16`).mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Remaining', '99');
            ctx.body = loadFixtureJSON('forecast_personal.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...subscribedClientOptions,
            baseEndpoint: url.toString(),
        });

        const subscribedForecast = await instance.estimate();
        expect(subscribedForecast).toMatchSnapshot('snapshot2');
        expect(subscribedForecast.timezone).toBe('Europe/Amsterdam');
        expect(subscribedForecast.accountType).toBe(AccountType.Personal);

        expect(subscribedForecast.energyProductionToday).toBe(5788);
        expect(subscribedForecast.energyProductionTomorrow).toBe(7507);

        expect(subscribedForecast.powerProductionNow).toBe(92);
        expect(subscribedForecast.energyProductionTodayRemaining).toBe(5783);
        expect(subscribedForecast.energyCurrentHour).toBe(96);

        expect(subscribedForecast.powerHighestPeakTimeToday).toBeInstanceOf(moment);
        expect(subscribedForecast.powerHighestPeakTimeToday).toStrictEqual(moment('2024-04-27T13:30:00+02:00'));
        expect(subscribedForecast.powerHighestPeakTimeTomorrow).toBeInstanceOf(moment);

        expect(subscribedForecast.powerHighestPeakTimeTomorrow).toStrictEqual(moment(
            '2024-04-28T14:30:00+02:00',
        ));

        expect(subscribedForecast.sumEnergyProduction(1)).toBe(216);
        expect(subscribedForecast.sumEnergyProduction(6)).toBe(2802);
        expect(subscribedForecast.sumEnergyProduction(12)).toBe(5582);
        expect(subscribedForecast.sumEnergyProduction(24)).toBe(5784);

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should return a forecast with professional type', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-04-27T07:00:00+02:00'));

        const route = mockServer.get(`/${subscribedClientOptions.apiKey}/estimate/52.16/4.47/20/10/2.16`).mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Remaining', '99');
            ctx.body = loadFixtureJSON('forecast_professional.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...subscribedClientOptions,
            baseEndpoint: url.toString(),
        });

        const subscribedForecast = await instance.estimate();
        expect(subscribedForecast).toMatchSnapshot('snapshot3');
        expect(subscribedForecast.timezone).toBe('Europe/Amsterdam');
        expect(subscribedForecast.accountType).toBe(AccountType.Professional);

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should raise an error with invalid data', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-06-11T12:00:00+02:00'));

        const route = mockServer.get('/estimate/48.21/16.36/23.44/180/5').mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Remaining', '99');
            ctx.body = loadFixtureJSON('invalid_forecast.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        await expect(instance.estimate()).rejects.toThrowErrorMatchingSnapshot('invalid_data');

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should raise an error because headers are missing', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-06-11T12:00:00+02:00'));

        const route = mockServer.get('/estimate/48.21/16.36/23.44/180/5').mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.body = loadFixtureJSON('invalid_forecast.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        await expect(instance.estimate()).rejects.toThrowErrorMatchingSnapshot('headers_missing');

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should parse the retry-at header', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-06-11T12:00:00+02:00'));

        const route = mockServer.get('/estimate/48.21/16.36/23.44/180/5').mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Retry-At', moment().add(10, 'hours').toISOString());
            ctx.body = loadFixtureJSON('forecast.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        const result = await instance.estimate();

        expect(result.apiRateLimit).toBe(12);

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should raise an error because invalid date', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-06-11T12:00:00+02:00'));

        const route = mockServer.get('/estimate/48.21/16.36/23.44/180/5').mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Remaining', '99');
            ctx.body = loadFixtureJSON('forecast.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        const invalidDay = moment('2026-06-11T12:00:00+02:00');

        const result = await instance.estimate();

        expect(() => result.dayProduction(invalidDay)).toThrowErrorMatchingSnapshot('invalid_date');

        expect(route).toHaveBeenCalledTimes(1);
    });

    it('should raise an error because no time found', async () => {
        jest.useFakeTimers()
            .setSystemTime(new Date('2024-06-11T12:00:00+02:00'));

        const route = mockServer.get('/estimate/48.21/16.36/23.44/180/5').mockImplementationOnce((ctx) => {
            ctx.status = 200;
            ctx.set('X-Ratelimit-Limit', '100');
            ctx.set('X-Ratelimit-Period', '60');
            ctx.set('X-Ratelimit-Remaining', '99');
            ctx.body = loadFixtureJSON('forecast.jsonc');
        });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        const invalidDay = moment('2026-06-11T12:00:00+02:00');

        const result = await instance.estimate();

        expect(() => result.peakProductionTime(invalidDay)).toThrowErrorMatchingSnapshot('no_time_found');

        expect(route).toHaveBeenCalledTimes(1);
    });
});

describe('Test api key validation', () => {
    const mockServer: MockServer = new MockServer();

    beforeAll(async () => {
        await mockServer.start();
    });

    afterAll(async () => {
        await mockServer.stop();
    });

    beforeEach(() => {
        mockServer.reset();
    });

    it('should return true when the API key is valid', async () => {
        const route = mockServer.get(`/${subscribedClientOptions.apiKey}/info`)
            .mockImplementationOnce((ctx) => {
                ctx.status = 200;
                ctx.set('X-Ratelimit-Limit', '100');
                ctx.set('X-Ratelimit-Period', '60');
                ctx.set('X-Ratelimit-Remaining', '99');
                ctx.body = loadFixtureJSON('validate_key.jsonc');
            });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...subscribedClientOptions,
            baseEndpoint: url.toString(),
        });

        await expect(instance.validateApiKey()).resolves.toBe(true);

        expect(route).toHaveBeenCalledTimes(1);
    });
});

describe('Test plane validation', () => {
    const mockServer: MockServer = new MockServer();

    beforeAll(async () => {
        await mockServer.start();
    });

    afterAll(async () => {
        await mockServer.stop();
    });

    beforeEach(() => {
        mockServer.reset();
    });

    it('should return true when the plane is valid', async () => {
        const route = mockServer.get('/check/48.21/16.36/23.44/180/5')
            .mockImplementationOnce((ctx) => {
                ctx.status = 200;
                ctx.set('X-Ratelimit-Limit', '100');
                ctx.set('X-Ratelimit-Period', '60');
                ctx.set('X-Ratelimit-Remaining', '99');
                ctx.body = loadFixtureJSON('validate_plane.jsonc');
            });

        const url = mockServer.getURL();

        const instance = new ForecastSolar({
            ...clientOptions,
            baseEndpoint: url.toString(),
        });

        await expect(instance.validatePlane()).resolves.toBe(true);

        expect(route).toHaveBeenCalledTimes(1);
    });
});
