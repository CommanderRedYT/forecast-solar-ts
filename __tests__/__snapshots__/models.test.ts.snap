// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test estimated forecast should returm a forecast with subscription: snapshot2 1`] = `
Estimate {
  "apiRateLimit": 60,
  "apiTimezone": "Europe/Amsterdam",
  "watts": {
    "2024-04-27T06:18:15+02:00": 0,
    "2024-04-27T06:30:00+02:00": 47,
    "2024-04-27T07:00:00+02:00": 92,
    "2024-04-27T07:30:00+02:00": 153,
    "2024-04-27T08:00:00+02:00": 215,
    "2024-04-27T08:30:00+02:00": 279,
    "2024-04-27T09:00:00+02:00": 342,
    "2024-04-27T09:30:00+02:00": 402,
    "2024-04-27T10:00:00+02:00": 457,
    "2024-04-27T10:30:00+02:00": 501,
    "2024-04-27T11:00:00+02:00": 545,
    "2024-04-27T11:30:00+02:00": 587,
    "2024-04-27T12:00:00+02:00": 616,
    "2024-04-27T12:30:00+02:00": 625,
    "2024-04-27T13:00:00+02:00": 635,
    "2024-04-27T13:30:00+02:00": 639,
    "2024-04-27T14:00:00+02:00": 635,
    "2024-04-27T14:30:00+02:00": 619,
    "2024-04-27T15:00:00+02:00": 600,
    "2024-04-27T15:30:00+02:00": 570,
    "2024-04-27T16:00:00+02:00": 536,
    "2024-04-27T16:30:00+02:00": 500,
    "2024-04-27T17:00:00+02:00": 458,
    "2024-04-27T17:30:00+02:00": 403,
    "2024-04-27T18:00:00+02:00": 344,
    "2024-04-27T18:30:00+02:00": 281,
    "2024-04-27T19:00:00+02:00": 215,
    "2024-04-27T19:30:00+02:00": 147,
    "2024-04-27T20:00:00+02:00": 85,
    "2024-04-27T20:30:00+02:00": 42,
    "2024-04-27T21:00:00+02:00": 18,
    "2024-04-27T21:01:07+02:00": 0,
    "2024-04-28T06:16:16+02:00": 0,
    "2024-04-28T06:30:00+02:00": 45,
    "2024-04-28T07:00:00+02:00": 88,
    "2024-04-28T07:30:00+02:00": 147,
    "2024-04-28T08:00:00+02:00": 208,
    "2024-04-28T08:30:00+02:00": 270,
    "2024-04-28T09:00:00+02:00": 332,
    "2024-04-28T09:30:00+02:00": 396,
    "2024-04-28T10:00:00+02:00": 487,
    "2024-04-28T10:30:00+02:00": 636,
    "2024-04-28T11:00:00+02:00": 778,
    "2024-04-28T11:30:00+02:00": 822,
    "2024-04-28T12:00:00+02:00": 801,
    "2024-04-28T12:30:00+02:00": 751,
    "2024-04-28T13:00:00+02:00": 746,
    "2024-04-28T13:30:00+02:00": 787,
    "2024-04-28T14:00:00+02:00": 837,
    "2024-04-28T14:30:00+02:00": 865,
    "2024-04-28T15:00:00+02:00": 855,
    "2024-04-28T15:30:00+02:00": 804,
    "2024-04-28T16:00:00+02:00": 765,
    "2024-04-28T16:30:00+02:00": 739,
    "2024-04-28T17:00:00+02:00": 687,
    "2024-04-28T17:30:00+02:00": 594,
    "2024-04-28T18:00:00+02:00": 500,
    "2024-04-28T18:30:00+02:00": 408,
    "2024-04-28T19:00:00+02:00": 309,
    "2024-04-28T19:30:00+02:00": 201,
    "2024-04-28T20:00:00+02:00": 103,
    "2024-04-28T20:30:00+02:00": 47,
    "2024-04-28T21:00:00+02:00": 19,
    "2024-04-28T21:02:49+02:00": 0,
    "2024-04-29T06:14:18+02:00": 0,
    "2024-04-29T06:30:00+02:00": 41,
    "2024-04-29T07:00:00+02:00": 93,
    "2024-04-29T07:30:00+02:00": 184,
    "2024-04-29T08:00:00+02:00": 291,
    "2024-04-29T08:30:00+02:00": 426,
    "2024-04-29T09:00:00+02:00": 569,
    "2024-04-29T09:30:00+02:00": 695,
    "2024-04-29T10:00:00+02:00": 735,
    "2024-04-29T10:30:00+02:00": 691,
    "2024-04-29T11:00:00+02:00": 707,
    "2024-04-29T11:30:00+02:00": 815,
    "2024-04-29T12:00:00+02:00": 910,
    "2024-04-29T12:30:00+02:00": 935,
    "2024-04-29T13:00:00+02:00": 919,
    "2024-04-29T13:30:00+02:00": 854,
    "2024-04-29T14:00:00+02:00": 768,
    "2024-04-29T14:30:00+02:00": 683,
    "2024-04-29T15:00:00+02:00": 628,
    "2024-04-29T15:30:00+02:00": 585,
    "2024-04-29T16:00:00+02:00": 544,
    "2024-04-29T16:30:00+02:00": 499,
    "2024-04-29T17:00:00+02:00": 452,
    "2024-04-29T17:30:00+02:00": 398,
    "2024-04-29T18:00:00+02:00": 339,
    "2024-04-29T18:30:00+02:00": 275,
    "2024-04-29T19:00:00+02:00": 208,
    "2024-04-29T19:30:00+02:00": 140,
    "2024-04-29T20:00:00+02:00": 79,
    "2024-04-29T20:30:00+02:00": 39,
    "2024-04-29T21:00:00+02:00": 16,
    "2024-04-29T21:04:31+02:00": 0,
    "2024-04-30T06:12:21+02:00": 0,
    "2024-04-30T06:30:00+02:00": 41,
    "2024-04-30T07:00:00+02:00": 82,
    "2024-04-30T07:30:00+02:00": 141,
    "2024-04-30T08:00:00+02:00": 203,
    "2024-04-30T08:30:00+02:00": 266,
    "2024-04-30T09:00:00+02:00": 329,
    "2024-04-30T09:30:00+02:00": 390,
    "2024-04-30T10:00:00+02:00": 445,
    "2024-04-30T10:30:00+02:00": 490,
    "2024-04-30T11:00:00+02:00": 531,
    "2024-04-30T11:30:00+02:00": 566,
    "2024-04-30T12:00:00+02:00": 594,
    "2024-04-30T12:30:00+02:00": 611,
    "2024-04-30T13:00:00+02:00": 623,
    "2024-04-30T13:30:00+02:00": 625,
    "2024-04-30T14:00:00+02:00": 623,
    "2024-04-30T14:30:00+02:00": 615,
    "2024-04-30T15:00:00+02:00": 599,
    "2024-04-30T15:30:00+02:00": 570,
    "2024-04-30T16:00:00+02:00": 535,
    "2024-04-30T16:30:00+02:00": 496,
    "2024-04-30T17:00:00+02:00": 452,
    "2024-04-30T17:30:00+02:00": 401,
    "2024-04-30T18:00:00+02:00": 343,
    "2024-04-30T18:30:00+02:00": 278,
    "2024-04-30T19:00:00+02:00": 209,
    "2024-04-30T19:30:00+02:00": 138,
    "2024-04-30T20:00:00+02:00": 76,
    "2024-04-30T20:30:00+02:00": 37,
    "2024-04-30T21:00:00+02:00": 15,
    "2024-04-30T21:06:12+02:00": 0,
  },
  "whDays": {
    "2024-04-27": 5788,
    "2024-04-28": 7507,
    "2024-04-29": 7256,
    "2024-04-30": 5657,
  },
  "whPeriod": {
    "2024-04-27T06:18:15+02:00": 0,
    "2024-04-27T06:30:00+02:00": 5,
    "2024-04-27T07:00:00+02:00": 35,
    "2024-04-27T07:30:00+02:00": 61,
    "2024-04-27T08:00:00+02:00": 92,
    "2024-04-27T08:30:00+02:00": 124,
    "2024-04-27T09:00:00+02:00": 155,
    "2024-04-27T09:30:00+02:00": 186,
    "2024-04-27T10:00:00+02:00": 215,
    "2024-04-27T10:30:00+02:00": 240,
    "2024-04-27T11:00:00+02:00": 262,
    "2024-04-27T11:30:00+02:00": 283,
    "2024-04-27T12:00:00+02:00": 301,
    "2024-04-27T12:30:00+02:00": 310,
    "2024-04-27T13:00:00+02:00": 315,
    "2024-04-27T13:30:00+02:00": 319,
    "2024-04-27T14:00:00+02:00": 319,
    "2024-04-27T14:30:00+02:00": 314,
    "2024-04-27T15:00:00+02:00": 305,
    "2024-04-27T15:30:00+02:00": 293,
    "2024-04-27T16:00:00+02:00": 277,
    "2024-04-27T16:30:00+02:00": 259,
    "2024-04-27T17:00:00+02:00": 240,
    "2024-04-27T17:30:00+02:00": 215,
    "2024-04-27T18:00:00+02:00": 187,
    "2024-04-27T18:30:00+02:00": 156,
    "2024-04-27T19:00:00+02:00": 124,
    "2024-04-27T19:30:00+02:00": 91,
    "2024-04-27T20:00:00+02:00": 58,
    "2024-04-27T20:30:00+02:00": 32,
    "2024-04-27T21:00:00+02:00": 15,
    "2024-04-27T21:01:07+02:00": 0,
    "2024-04-28T06:16:16+02:00": 0,
    "2024-04-28T06:30:00+02:00": 5,
    "2024-04-28T07:00:00+02:00": 33,
    "2024-04-28T07:30:00+02:00": 59,
    "2024-04-28T08:00:00+02:00": 89,
    "2024-04-28T08:30:00+02:00": 120,
    "2024-04-28T09:00:00+02:00": 151,
    "2024-04-28T09:30:00+02:00": 182,
    "2024-04-28T10:00:00+02:00": 221,
    "2024-04-28T10:30:00+02:00": 281,
    "2024-04-28T11:00:00+02:00": 354,
    "2024-04-28T11:30:00+02:00": 400,
    "2024-04-28T12:00:00+02:00": 406,
    "2024-04-28T12:30:00+02:00": 388,
    "2024-04-28T13:00:00+02:00": 374,
    "2024-04-28T13:30:00+02:00": 383,
    "2024-04-28T14:00:00+02:00": 406,
    "2024-04-28T14:30:00+02:00": 426,
    "2024-04-28T15:00:00+02:00": 430,
    "2024-04-28T15:30:00+02:00": 415,
    "2024-04-28T16:00:00+02:00": 392,
    "2024-04-28T16:30:00+02:00": 376,
    "2024-04-28T17:00:00+02:00": 357,
    "2024-04-28T17:30:00+02:00": 320,
    "2024-04-28T18:00:00+02:00": 274,
    "2024-04-28T18:30:00+02:00": 227,
    "2024-04-28T19:00:00+02:00": 179,
    "2024-04-28T19:30:00+02:00": 128,
    "2024-04-28T20:00:00+02:00": 76,
    "2024-04-28T20:30:00+02:00": 38,
    "2024-04-28T21:00:00+02:00": 17,
    "2024-04-28T21:02:49+02:00": 0,
    "2024-04-29T06:14:18+02:00": 0,
    "2024-04-29T06:30:00+02:00": 5,
    "2024-04-29T07:00:00+02:00": 34,
    "2024-04-29T07:30:00+02:00": 69,
    "2024-04-29T08:00:00+02:00": 119,
    "2024-04-29T08:30:00+02:00": 179,
    "2024-04-29T09:00:00+02:00": 249,
    "2024-04-29T09:30:00+02:00": 316,
    "2024-04-29T10:00:00+02:00": 358,
    "2024-04-29T10:30:00+02:00": 357,
    "2024-04-29T11:00:00+02:00": 350,
    "2024-04-29T11:30:00+02:00": 381,
    "2024-04-29T12:00:00+02:00": 431,
    "2024-04-29T12:30:00+02:00": 461,
    "2024-04-29T13:00:00+02:00": 464,
    "2024-04-29T13:30:00+02:00": 443,
    "2024-04-29T14:00:00+02:00": 406,
    "2024-04-29T14:30:00+02:00": 363,
    "2024-04-29T15:00:00+02:00": 328,
    "2024-04-29T15:30:00+02:00": 303,
    "2024-04-29T16:00:00+02:00": 282,
    "2024-04-29T16:30:00+02:00": 261,
    "2024-04-29T17:00:00+02:00": 238,
    "2024-04-29T17:30:00+02:00": 213,
    "2024-04-29T18:00:00+02:00": 184,
    "2024-04-29T18:30:00+02:00": 154,
    "2024-04-29T19:00:00+02:00": 121,
    "2024-04-29T19:30:00+02:00": 87,
    "2024-04-29T20:00:00+02:00": 55,
    "2024-04-29T20:30:00+02:00": 30,
    "2024-04-29T21:00:00+02:00": 14,
    "2024-04-29T21:04:31+02:00": 1,
    "2024-04-30T06:12:21+02:00": 0,
    "2024-04-30T06:30:00+02:00": 6,
    "2024-04-30T07:00:00+02:00": 31,
    "2024-04-30T07:30:00+02:00": 56,
    "2024-04-30T08:00:00+02:00": 86,
    "2024-04-30T08:30:00+02:00": 117,
    "2024-04-30T09:00:00+02:00": 149,
    "2024-04-30T09:30:00+02:00": 180,
    "2024-04-30T10:00:00+02:00": 209,
    "2024-04-30T10:30:00+02:00": 234,
    "2024-04-30T11:00:00+02:00": 255,
    "2024-04-30T11:30:00+02:00": 274,
    "2024-04-30T12:00:00+02:00": 290,
    "2024-04-30T12:30:00+02:00": 301,
    "2024-04-30T13:00:00+02:00": 309,
    "2024-04-30T13:30:00+02:00": 312,
    "2024-04-30T14:00:00+02:00": 312,
    "2024-04-30T14:30:00+02:00": 310,
    "2024-04-30T15:00:00+02:00": 304,
    "2024-04-30T15:30:00+02:00": 292,
    "2024-04-30T16:00:00+02:00": 276,
    "2024-04-30T16:30:00+02:00": 258,
    "2024-04-30T17:00:00+02:00": 237,
    "2024-04-30T17:30:00+02:00": 213,
    "2024-04-30T18:00:00+02:00": 186,
    "2024-04-30T18:30:00+02:00": 155,
    "2024-04-30T19:00:00+02:00": 122,
    "2024-04-30T19:30:00+02:00": 87,
    "2024-04-30T20:00:00+02:00": 54,
    "2024-04-30T20:30:00+02:00": 28,
    "2024-04-30T21:00:00+02:00": 13,
    "2024-04-30T21:06:12+02:00": 1,
  },
}
`;

exports[`Test estimated forecast should return a forecast: snapshot1 1`] = `
Estimate {
  "apiRateLimit": 12,
  "apiTimezone": "Europe/Vienna",
  "watts": {
    "2024-06-11T04:53:46+02:00": 0,
    "2024-06-11T05:15:00+02:00": 18,
    "2024-06-11T06:00:00+02:00": 62,
    "2024-06-11T07:00:00+02:00": 120,
    "2024-06-11T08:00:00+02:00": 168,
    "2024-06-11T09:00:00+02:00": 229,
    "2024-06-11T10:00:00+02:00": 225,
    "2024-06-11T11:00:00+02:00": 250,
    "2024-06-11T12:00:00+02:00": 289,
    "2024-06-11T13:00:00+02:00": 335,
    "2024-06-11T14:00:00+02:00": 343,
    "2024-06-11T15:00:00+02:00": 311,
    "2024-06-11T16:00:00+02:00": 253,
    "2024-06-11T17:00:00+02:00": 182,
    "2024-06-11T18:00:00+02:00": 119,
    "2024-06-11T19:00:00+02:00": 68,
    "2024-06-11T20:00:00+02:00": 33,
    "2024-06-11T20:54:51+02:00": 0,
    "2024-06-12T04:53:36+02:00": 0,
    "2024-06-12T05:15:00+02:00": 18,
    "2024-06-12T06:00:00+02:00": 53,
    "2024-06-12T07:00:00+02:00": 101,
    "2024-06-12T08:00:00+02:00": 148,
    "2024-06-12T09:00:00+02:00": 190,
    "2024-06-12T10:00:00+02:00": 223,
    "2024-06-12T11:00:00+02:00": 250,
    "2024-06-12T12:00:00+02:00": 268,
    "2024-06-12T13:00:00+02:00": 276,
    "2024-06-12T14:00:00+02:00": 264,
    "2024-06-12T15:00:00+02:00": 238,
    "2024-06-12T16:00:00+02:00": 204,
    "2024-06-12T17:00:00+02:00": 164,
    "2024-06-12T18:00:00+02:00": 118,
    "2024-06-12T19:00:00+02:00": 68,
    "2024-06-12T20:00:00+02:00": 33,
    "2024-06-12T20:55:26+02:00": 0,
  },
  "whDays": {
    "2024-06-11": 2991,
    "2024-06-12": 2602,
  },
  "whPeriod": {
    "2024-06-11T04:53:46+02:00": 0,
    "2024-06-11T05:15:00+02:00": 3,
    "2024-06-11T06:00:00+02:00": 30,
    "2024-06-11T07:00:00+02:00": 91,
    "2024-06-11T08:00:00+02:00": 144,
    "2024-06-11T09:00:00+02:00": 199,
    "2024-06-11T10:00:00+02:00": 227,
    "2024-06-11T11:00:00+02:00": 238,
    "2024-06-11T12:00:00+02:00": 270,
    "2024-06-11T13:00:00+02:00": 312,
    "2024-06-11T14:00:00+02:00": 339,
    "2024-06-11T15:00:00+02:00": 327,
    "2024-06-11T16:00:00+02:00": 282,
    "2024-06-11T17:00:00+02:00": 218,
    "2024-06-11T18:00:00+02:00": 151,
    "2024-06-11T19:00:00+02:00": 94,
    "2024-06-11T20:00:00+02:00": 51,
    "2024-06-11T20:54:51+02:00": 15,
    "2024-06-12T04:53:36+02:00": 0,
    "2024-06-12T05:15:00+02:00": 3,
    "2024-06-12T06:00:00+02:00": 27,
    "2024-06-12T07:00:00+02:00": 77,
    "2024-06-12T08:00:00+02:00": 125,
    "2024-06-12T09:00:00+02:00": 169,
    "2024-06-12T10:00:00+02:00": 207,
    "2024-06-12T11:00:00+02:00": 237,
    "2024-06-12T12:00:00+02:00": 259,
    "2024-06-12T13:00:00+02:00": 272,
    "2024-06-12T14:00:00+02:00": 270,
    "2024-06-12T15:00:00+02:00": 251,
    "2024-06-12T16:00:00+02:00": 221,
    "2024-06-12T17:00:00+02:00": 184,
    "2024-06-12T18:00:00+02:00": 141,
    "2024-06-12T19:00:00+02:00": 93,
    "2024-06-12T20:00:00+02:00": 51,
    "2024-06-12T20:55:26+02:00": 15,
  },
}
`;
